### PR TITLE
Handle endianness

### PR DESCRIPTION
**Pull request by davidbrochart**

### CODE REVIEW

File: include/xtensor-io/xio_binary.hpp [hunk 1]
```diff
@@ -18,28 +18,69 @@ namespace xt
...
             auto shape = eval_ex.shape();
-            std::size_t uncompressed_size = compute_size(shape) * sizeof(value_type);
+            std::size_t size = compute_size(shape);
+            std::size_t uncompressed_size = size * sizeof(value_type);
             const char* uncompressed_buffer = reinterpret_cast<const char*>(eval_ex.data());
-            stream.write(uncompressed_buffer, std::streamsize(uncompressed_size));
-            stream.flush();
+            if ((sizeof(value_type) > 1) && (as_big_endian != is_big_endian()))
+            {
+                char* swapped_buffer = new char[size * sizeof(value_type)];
```
Comment (JohanMabille) [inline include/xtensor-io/xio_binary.hpp:73]: Please use `std::vector<char>` instead, it manages the memory for you and avoids memory leak if an exception is thrown.

File: include/xtensor-io/xio_binary.hpp [hunk 2]
```diff
@@ -18,28 +18,69 @@ namespace xt
...
+        {
+            for (std::size_t i = 0; i < size; i++)
+            {
+                T tmp = *reinterpret_cast<const T*>(&(in_buffer[i * sizeof(T)]));
+                for (std::size_t j = 0; j < sizeof(T); j++)
+                {
+                    out_buffer[i * sizeof(T) + j] = reinterpret_cast<char*>(&tmp)[(i + 1) * sizeof(T) - j - 1];
+                }
+            }
+        }
```
Comment (JohanMabille) [inline include/xtensor-io/xio_binary.hpp:42]: This couls be simplified in
```cpp
tempate <class T>
void swap_endianness(xt::svector<T>& buffer)
{
    char* buf = reinterpret_cast<char*>(buffer.data());
    char* end  = buf + buffer.size() * sizeof(T);
    while(buf != end)
    {
        std::reverse(buf, buf + sizeof(T));
        buf += sizeof(T);
    }
```

File: include/xtensor-io/xio_binary.hpp [hunk 3]
```diff
@@ -18,28 +18,69 @@ namespace xt
...
             stream.seekg(0, stream.beg);
-            xt::svector<T> uncompressed_buffer(uncompressed_size / sizeof(T));
-            stream.read(reinterpret_cast<char*>(uncompressed_buffer.data()), (std::streamsize)uncompressed_size);
+            std::size_t size = uncompressed_size / sizeof(T);
+            xt::svector<T> uncompressed_buffer(size);
+            char* buffer = reinterpret_cast<char*>(uncompressed_buffer.data());
+            stream.read(buffer, (std::streamsize)uncompressed_size);
+            if ((sizeof(T) > 1) && (as_big_endian != is_big_endian()))
+            {
+                swap_endianness<T>(buffer, buffer, size);
```
Comment (JohanMabille) [inline include/xtensor-io/xio_binary.hpp:56]: With the simplification of `swap_endianness` this simplifies into:
```cpp
swap_endianness(uncompressed_buffer);
```

**Review by JohanMabille — CHANGES_REQUESTED** [review]

**Comment by davidbrochart** [comment]
Thanks for the review @JohanMabille, I will take your comments into account.
First I'd like to discuss if it's the right place to handle endianness - in the load/dump functions of each storage formats. The drawback is that each of them has to support endianness.
On the other hand, we can't just have a general function that swaps endianness of an array data, because e.g. in `xfile_array` we don't have control over when the array is loaded or dumped. But we could imagine having a callback before dumping and after loading?

File: include/xtensor-io/xio_binary.hpp (hunk 2 above)
Comment (davidbrochart) [reply include/xtensor-io/xio_binary.hpp:41]: Great simplification!

**Comment by davidbrochart** [comment]
@JohanMabille If you think it's fine, I will make the change to `xblosc` and `xgzip` also.

**Review by JohanMabille — APPROVED** [review]

**Comment by JohanMabille** [comment]
I think it's the right place to handle endianness. However, the `as_big_endian` parameter could be defaulted to `is_big_endian` so that formats that don't want to handle endianness don't have to do so.

**Comment by davidbrochart** [comment]
Good point.
Should we move `is_big_endian()` and `swap_endianness()` to `xtensor-io.hpp`, so that it's available to every format?

**Comment by JohanMabille** [comment]
I don't think so, endianness is relevant to binary formats only.

**Comment by davidbrochart** [comment]
So `xblosc` and `xgzip` will need it, right?

**Comment by JohanMabille** [comment]
Ah my bad, I thought `xio_binary.hpp` was included in other files as a basic buiding block. Therefore, indeed it would make sense ot move the functions ton `xtensor_io.hpp`, sorry for the confusion!

**Pull request merged**